python/ray/serve/_private/router.py

```diff
@@ -373,13 +508,21 @@ async def choose_replica_for_query(self, query: Query) -> ReplicaWrapper:
         Upon cancellation (by the caller), the future is cancelled and will be passed
         over when a replica becomes available.
         """
-        fut = asyncio.Future()
-        self._pending_assignment_futures.append(fut)
+        pending_request = PendingRequest(asyncio.Future(), query.metadata)
         try:
+            self._pending_requests_to_fulfill.append(pending_request)
+            self._pending_requests_to_schedule.append(pending_request)
             self.maybe_start_scheduling_tasks()
-            replica = await fut
-        except asyncio.CancelledError:
-            fut.cancel()
+            replica = await pending_request.future
+        except asyncio.CancelledError as e:
+            pending_request.future.cancel()
+            try:
+                self._pending_requests_to_fulfill.remove(pending_request)
```

Comment:
Why do we remove this here instead of at `pop` time? Wouldn't it be more efficient to lazily remove it to avoid shifting the whole queue?

Reply:
you're right, I think this was left from a previous WIP state where I was doing something slightly different. this should be safe

Reply:
updated